src/unix.rs

```diff
@@ -17,17 +17,26 @@ use std::thread::{self, Builder, JoinHandle};
 use std::time::Duration;
 
 #[derive(Debug)]
-pub enum Client {
-    /// `--jobserver-auth=R,W`
-    Pipe { read: File, write: File },
-    /// `--jobserver-auth=fifo:PATH`
-    Fifo {
-        file: File,
-        path: PathBuf,
-        /// it can only go from false -> true but not the other way around, since that
-        ///  could cause a race condition.
-        is_non_blocking: AtomicBool,
-    },
+/// This preserves the `--jobserver-auth` type at creation time,
+/// so auth type will be passed down to and inherit from sub-Make processes correctly.
+///
+/// See <https://github.com/rust-lang/jobserver-rs/issues/99> for details.
+enum ClientCreationArg {
+    Fds { read: c_int, write: c_int },
+    Fifo(Box<Path>),
+}
+
+#[derive(Debug)]
+pub struct Client {
+    read: File,
+    write: File,
+    creation_arg: ClientCreationArg,
+    /// It is set to `None` if the pipe is shared with other processesdoes so it cannot support
+    ///  non-blocking mode.
+    ///
+    /// If it is set to `Some`, then It can only go from `Some(false)` -> `Some(true)` but not
+    ///  the other way around, since that could cause a race condition.
```

Comment:
```suggestion
    /// It is set to `None` if the pipe is shared with other processes so it cannot support
    /// non-blocking mode.
    ///
    /// If it is set to `Some`, then it can only go from `Some(false)` -> `Some(true)` but not
    /// the other way around, since that could cause a race condition.
```